MdeModulePkg/Core/Dxe/Misc/MemoryProtectionSupport.c

```diff
@@ -0,0 +1,2853 @@
+       SpecialRegionEntryLink = SpecialRegionEntryLink->ForwardLink)
+  {
+    SpecialRegionEntry = CR (
+                           SpecialRegionEntryLink,
+                           MEMORY_PROTECTION_SPECIAL_REGION_LIST_ENTRY,
+                           Link,
+                           MEMORY_PROTECTION_SPECIAL_REGION_LIST_ENTRY_SIGNATURE
+                           );
+
+    DEBUG ((
+      DEBUG_INFO,
+      "Memory Protection Special Region: 0x%llx - 0x%llx. Attributes: 0x%llx\n",
+      SpecialRegionEntry->SpecialRegion.Start,
+      SpecialRegionEntry->SpecialRegion.Start + SpecialRegionEntry->SpecialRegion.Length,
+      SpecialRegionEntry->SpecialRegion.EfiAttributes
+      ));
+  }
+
+  return;
+}
+
+// ---------------------------------------
+//        GCD MEMORY MAP FUNCTIONS
+// ---------------------------------------
+
+/**
+  Sort the GCD memory map entries from low to high.
+
+  @param[in, out]   MemoryMap       A pointer to the buffer containing the current memory map
+  @param[in]        MemoryMapSize   Size, in bytes, of the MemoryMap buffer
+  @param[in]        DescriptorSize  Size, in bytes, of an individual EFI_GCD_MEMORY_SPACE_DESCRIPTOR
+**/
+STATIC
+VOID
+SortMemorySpaceMap (
+  IN OUT EFI_GCD_MEMORY_SPACE_DESCRIPTOR  *MemoryMap,
+  IN CONST UINTN                          *MemoryMapSize,
+  IN CONST UINTN                          *DescriptorSize
+  )
+{
+  EFI_GCD_MEMORY_SPACE_DESCRIPTOR  *MemoryMapEntry;
+  EFI_GCD_MEMORY_SPACE_DESCRIPTOR  *NextMemoryMapEntry;
+  EFI_GCD_MEMORY_SPACE_DESCRIPTOR  *MemoryMapEnd;
+  EFI_GCD_MEMORY_SPACE_DESCRIPTOR  TempMemoryMap;
+
+  if ((MemoryMap == NULL) || (MemoryMapSize == NULL) || (DescriptorSize == NULL)) {
+    return;
+  }
+
+  MemoryMapEntry     = MemoryMap;
+  NextMemoryMapEntry = NEXT_MEMORY_SPACE_DESCRIPTOR (MemoryMapEntry, *DescriptorSize);
+  MemoryMapEnd       = (EFI_GCD_MEMORY_SPACE_DESCRIPTOR *)((UINT8 *)MemoryMap + *MemoryMapSize);
+  while (MemoryMapEntry < MemoryMapEnd) {
+    while (NextMemoryMapEntry < MemoryMapEnd) {
+      if (MemoryMapEntry->BaseAddress > NextMemoryMapEntry->BaseAddress) {
+        CopyMem (&TempMemoryMap, MemoryMapEntry, *DescriptorSize);
+        CopyMem (MemoryMapEntry, NextMemoryMapEntry, *DescriptorSize);
+        CopyMem (NextMemoryMapEntry, &TempMemoryMap, *DescriptorSize);
+      }
+
+      NextMemoryMapEntry = NEXT_MEMORY_SPACE_DESCRIPTOR (NextMemoryMapEntry, *DescriptorSize);
+    }
+
+    MemoryMapEntry     = NEXT_MEMORY_SPACE_DESCRIPTOR (MemoryMapEntry, *DescriptorSize);
+    NextMemoryMapEntry = NEXT_MEMORY_SPACE_DESCRIPTOR (MemoryMapEntry, *DescriptorSize);
+  }
+
+  return;
+}
+
+/**
+  Convert a GCD memory type to an EFI memory type
+
+  @param[in]  GcdType   The GCD memory type to convert
+
+  @retval               The converted EFI memory type
+**/
+STATIC
+EFI_MEMORY_TYPE
+GcdTypeToEfiType (
+  IN EFI_GCD_MEMORY_TYPE  *GcdType
+  )
+{
+  if (GcdType == NULL) {

```

Comment:
Is the pointer really allowed to be `NULL`? If not, should there at least be an assert?

Reply:
No, there is only one caller of this function, it will not be null, and in fact there is no reason to pass a pointer at all, the variable is a local variable to that function. Will change to not be a pointer.